File: backend/src/domain/pricing.ts
import type { CartLine, Coupon, PaymentMethod, Product, Totals } from '../types.js';

export const FREE_SHIPPING_THRESHOLD = 499;
export const SHIPPING_FEE = 49;
export const COD_FEE = 30;

export interface CouponRule extends Coupon {
  firstOrderOnly?: boolean;
}

export class PricingError extends Error {}

/** Validates a coupon against the cart; returns a customer-facing reason when it cannot be used. */
export function couponProblem(c: CouponRule, sell: number, isFirstOrder: boolean): string | null {
  if (c.minOrder && sell < c.minOrder) return `${c.code} needs a minimum order of ₹${c.minOrder}.`;
  if (c.firstOrderOnly && !isFirstOrder) return `${c.code} is valid on your first order only.`;
  return null;
}

/**
 * Server-authoritative cart pricing. All prices are GST inclusive, so GST is
 * extracted from the selling price rather than added on top.
 */
export function quote(
  lines: CartLine[],
  catalog: Map<string, Product>,
  opts: { coupon?: CouponRule | null; payment?: PaymentMethod; isFirstOrder?: boolean } = {},
): Totals {
  const items = lines
    .filter((l) => l.qty > 0)
    .map((l) => {
      const p = catalog.get(l.sku);
      if (!p) throw new PricingError(`Unknown product ${l.sku}`);
      if (p.stock <= 0) throw new PricingError(`${p.name} is out of stock`);
      const qty = Math.min(Math.floor(l.qty), p.stock);
      return { sku: p.sku, qty, name: p.name, price: p.price, mrp: p.mrp, lineTotal: p.price * qty, gstPct: p.gst, concern: p.concern };
    });
  const mrp = items.reduce((a, i) => a + i.mrp * i.qty, 0);
  const sell = items.reduce((a, i) => a + i.lineTotal, 0);
  let shipping = sell >= FREE_SHIPPING_THRESHOLD || sell === 0 ? 0 : SHIPPING_FEE;
  let couponAmount = 0;
  const coupon = opts.coupon && !couponProblem(opts.coupon, sell, opts.isFirstOrder ?? false) ? opts.coupon : null;
  if (coupon) {
    if (coupon.ship) shipping = 0;
    else if (coupon.concern)
      couponAmount = Math.round(
        (items.filter((i) => i.concern === coupon.concern).reduce((a, i) => a + i.lineTotal, 0) * (coupon.pct ?? 0)) / 100,
      );
    else couponAmount = Math.round((sell * (coupon.pct ?? 0)) / 100);
  }
  const codFee = opts.payment === 'cod' && items.length ? COD_FEE : 0;
  const afterDiscounts = sell - couponAmount;
  const total = afterDiscounts + shipping + codFee;
  const gst = Math.round(items.reduce((a, i) => a + i.lineTotal * (i.gstPct / (100 + i.gstPct)), 0));
  const deliveryLabel =
    shipping === 0 ? (codFee ? `Free + ₹${COD_FEE} COD` : 'Free') : `₹${shipping}${codFee ? ` + ₹${COD_FEE} COD` : ''}`;
  return {
    items: items.map(({ sku, qty, name, price, mrp, lineTotal }) => ({ sku, qty, name, price, mrp, lineTotal })),
    mrp,
    discount: mrp - sell,
    coupon: couponAmount,
    couponCode: coupon?.code ?? null,
    afterDiscounts,
    shipping,
    codFee,
    deliveryLabel,
    total,
    gst,
  };
}

File: app/lib/data/pricing.dart
import 'models.dart';

// Same rules as backend/src/domain/pricing.ts. The app prices the bag locally
// for instant feedback; the backend re-prices from live Zoho data at checkout
// and that total is what the customer is charged.

const freeShippingThreshold = 499;
const shippingFee = 49;
const codFee = 30;

/// Returns a customer-facing reason the coupon can't be used, or null.
String? couponProblem(Coupon c, int sell, bool isFirstOrder) {
  if (c.minOrder > 0 && sell < c.minOrder) return '${c.code} needs a minimum order of ₹${c.minOrder}.';
  if (c.firstOrderOnly && !isFirstOrder) return '${c.code} is valid on your first order only.';
  return null;
}

Totals price(List<CartLine> lines, {Coupon? coupon, PaymentMethod? payment, bool isFirstOrder = true}) {
  final items = lines.where((l) => l.qty > 0).toList();
  final mrp = items.fold(0, (a, i) => a + i.product.mrp * i.qty);
  final sell = items.fold(0, (a, i) => a + i.lineTotal);
  var ship = sell >= freeShippingThreshold || sell == 0 ? 0 : shippingFee;
  var cp = 0;
  final valid = coupon != null && couponProblem(coupon, sell, isFirstOrder) == null ? coupon : null;
  if (valid != null) {
    if (valid.ship) {
      ship = 0;
    } else if (valid.concern != null) {
      cp = (items.where((i) => i.product.concern == valid.concern).fold(0, (a, i) => a + i.lineTotal) * valid.pct / 100).round();
    } else {
      cp = (sell * valid.pct / 100).round();
    }
  }
  final cod = payment == PaymentMethod.cod && items.isNotEmpty ? codFee : 0;
  final after = sell - cp;
  final gst = items.fold(0.0, (a, i) => a + i.lineTotal * (i.product.gst / (100 + i.product.gst))).round();
  final label = ship == 0 ? (cod > 0 ? 'Free + ₹$codFee COD' : 'Free') : '₹$ship${cod > 0 ? ' + ₹$codFee COD' : ''}';
  return Totals(
    items: items,
    mrp: mrp,
    discount: mrp - sell,
    coupon: cp,
    afterDiscounts: after,
    shipping: ship,
    codFee: cod,
    total: after + ship + cod,
    gst: gst,
    couponCode: valid?.code,
    deliveryLabel: label,
  );
}
